flang-rt/cmake/modules/AddFlangRT.cmake

```
@@ -231,6 +231,26 @@ function (add_flangrt_library name)
       target_compile_options(${tgtname} PRIVATE
           $<$<COMPILE_LANGUAGE:CXX>:-fno-exceptions -fno-rtti -funwind-tables -fno-asynchronous-unwind-tables>
         )
+
+      # We define our own _GLIBCXX_THROW_OR_ABORT here because, as of
+      # GCC 15.1, libstdc++ uses (void)_EXC in its definition in order
+      # to silence a warning.
+      #
+      # This is a problem for us because some compilers, specifically
+      # clang, do not always optimize away that (void)_EXC even though
+      # it is unreachable since it occurs after a call to
+      # _builtin_abort().  Because _EXC is typically an object derived
+      # from std::exception, (void)_EXC, when not optimized away,
+      # causes a link reference to std::exception, and we do not want
+      # to link against std::exception since doing that would link us
+      # against the C++ runtime library, and we do not want to link
+      # the Fortran runtime against the C++ runtime.
```

Reads like "causes A, and we do not want to do A since doing that would do B, and we do not want to do B" with A = "link to std::exception" and B = "link against C++ standard library". Can shorten and state the reason instead.

```suggestion
      # calls std::exception methods in the libstdc++ static or dynamic library. We cannot link libstdc++ because it may conflict with a C++ runtime of a hybrid Fortran/C++ application.
```

---

@Meinersbur I revised the comment in a way similar to your suggestion; please take a look.

Also, although I did mention .so version conflicts in the comment as the reason we don't want to link against libstdc++, are you sure that's really true?  The GNU .so files are forward compatible, so, as long as our users use a version at least as new as our build version, we should be fine there.  And, we do link the Fortran runtime against the GNU C runtime, which has the same forward-but-not-backward compatibility profile as the GNU C++ runtime.  So, I'm not sure .so conflicts would actually manifest.

My understanding was that we didn't want to link against libstdc++ because it's big and bulky and we just don't need it.

---

Some problems that could occur:
* flang-rt uses GCC libstdc++, application uses LLVM libc++ (or potentially another commercial runtime implementation)
* flang-rt statically links an older version of libstdc++.a, application links a newer version of libstdc++.so/.a

Generally, bad things happen if the linker or loader finds the same mangled symbol name with different implementations. There are solutions to this, like only supporting libstdc++.a and libflang_rt.runtime.so with `-fvisibility=hidden`, requiring the user to only link the same C++ standard library or newer version of it dynamically, or not depending on libstdc++.a/.so at all. We chose the latter, staying out of the symbol conflict game.

The size of libstdc++ is not a problem. When linking dynamically, libstdc++.so is present on the system anyway, and only touched pages are loaded into memory. When linking statically, the linker only takes the object files from libstdc++.a that are actually needed. C++ has a ["you only pay for what you use"](https://releases.llvm.org/3.1/docs/CodingStandards.html#ci_rtti_exceptions) principle. The two C++ features that violate this are exceptions and RTTI. Both are disabled for all of LLVM, including Flang-RT.